core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +668,954 @@ impl CheckedAdd for Fixed64 {
+						lhs > rhs
+					};
+					if predicate_1 || predicate_2 {
+						*qhat.borrow_mut() -= 1;
+						*rhat.borrow_mut() += other_norm.get(n - 1) as Double;
+						true
+					} else {
+						false
+					}
+				};
+
+				test();
+				while (*rhat.borrow() as Double) < B {
+					if !test() { break; }
+				}
+
+				let qhat = qhat.into_inner();
+				// we don't need rhat anymore. just let it go out of scope when it does.
+
+				// step D4
+				let lhs = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+				let rhs = other_norm.clone().mul(Self::from(qhat));
+
+				let maybe_sub = lhs.sub(rhs);
+				let mut negative = false;
+				let sub = match maybe_sub {
+					Ok(t) => t,
+					Err(t) => { negative = true; t }
+				};
+				(j..=j+n).for_each(|d| { self_norm.set(d, sub.get(d - j)); });
+
+				// step D5
+				// PROOF: the `test()` specifically decreases qhat until it is below `B`. conversion
+				// is safe.
+				q.set(j, qhat as Single);
+
+				// step D6: add back if negative happened.
+				if negative {
+					q.set(j, q.get(j) - 1);
+					let u = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+					let r = other_norm.clone().add(u);
+					(j..=j+n).rev().for_each(|d| { self_norm.set(d, r.get(d - j)); })
+				}
+			}
+
+			// if requested, calculate remainder.
+			if rem {
+				// undo the normalization.
+				if normalizer_bits > 0 {
+					let s = SHIFT as u32;
+					let nb = normalizer_bits;
+					for d in 0..n-1 {
+						let v = self_norm.get(d) >> nb
+							| self_norm.get(d + 1).overflowing_shl(s - nb).0;
+						r.set(d, v);
+					}
+					r.set(n - 1, self_norm.get(n - 1) >> normalizer_bits);
+				} else {
+					r = self_norm;
+				}
+			}
+
+			Some((q, r))
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl rstd::fmt::Debug for Number{
+		fn fmt(&self, f: &mut rstd::fmt::Formatter<'_>) -> rstd::fmt::Result {
+			write!(
+				f,
+				"Number {{ {:?} ({:?})}}",
+				self.digits,
+				u128::try_from(self.clone()).unwrap_or_else(|_| 0),
+			)
+		}
+	}
+
+	impl PartialEq for Number {
+		fn eq(&self, other: &Self) -> bool {
+			// sadly, we have to reallocate here as strip mutably uses self.
+			let mut lhs = self.clone();
+			let mut rhs = other.clone();
+			lhs.lstrip();
+			rhs.lstrip();
+			lhs.digits.eq(&rhs.digits)
+		}
+	}
+
+	impl Eq for Number {}
+
+	impl Ord for Number {
+		fn cmp(&self, other: &Self) -> Ordering {
+			let lhs_first = self.digits.iter().position(|&e| e != 0);
+			let rhs_first = other.digits.iter().position(|&e| e != 0);
+
+			match (lhs_first, rhs_first) {
+				// edge cases that should not happen. This basically means that one or both were
+				// zero.
+				(None, None) => Ordering::Equal,
+				(Some(_), None) => Ordering::Greater,
+				(None, Some(_)) => Ordering::Less,
+				(Some(lhs_idx), Some(rhs_idx)) => {
+					let lhs = &self.digits[lhs_idx..];
+					let rhs = &other.digits[rhs_idx..];
+					let len_cmp = lhs.len().cmp(&rhs.len());
+					match len_cmp {
```

Comment:
matching on only one arm can be done with `if let`, but OK anyway

Reply:
If you are not super strict on it, I find match to be more readable. Surprisingly, on the current version match is actually 1 line of code shorter than if/else.